presto-main/src/main/java/com/facebook/presto/execution/scheduler/TableWriteInfo.java

```diff
@@ -0,0 +1,202 @@
+            }
+            if (target instanceof TableWriterNode.InsertReference) {
+                TableWriterNode.InsertReference insert = (TableWriterNode.InsertReference) target;
+                return Optional.of(new ExecutionWriterTarget.InsertHandle(metadata.beginInsert(session, insert.getHandle()), metadata.getTableMetadata(session, insert.getHandle()).getTable()));
+            }
+            if (target instanceof TableWriterNode.DeleteHandle) {
+                TableWriterNode.DeleteHandle delete = (TableWriterNode.DeleteHandle) target;
+                return Optional.of(new ExecutionWriterTarget.DeleteHandle(metadata.beginDelete(session, delete.getHandle()), delete.getSchemaTableName()));
+            }
+            throw new IllegalArgumentException("Unhandled target type: " + target.getClass().getSimpleName());
+        }
+
+        return Optional.empty();
+    }
+
+    private static Optional<AnalyzeTableHandle> createAnalyzeTableHandle(StreamingSubPlan plan, Metadata metadata, Session session)
+    {
+        Optional<StatisticsWriterNode> node = findSinglePlanNode(plan, StatisticsWriterNode.class);
+        if (node.isPresent()) {
+            return Optional.of(metadata.beginStatisticsCollection(session, ((StatisticsWriterNode.WriteStatisticsReference) node.get().getTarget()).getHandle()));
+        }
+        return Optional.empty();
+    }
+
+    private static Optional<TableHandle> createDeleteTableHandle(StreamingSubPlan plan, Optional<ExecutionWriterTarget> writerTarget, Metadata metadata, Session session)
+    {
+        if (writerTarget.isPresent() && writerTarget.get() instanceof ExecutionWriterTarget.DeleteHandle) {
+            TableHandle tableHandle = ((ExecutionWriterTarget.DeleteHandle) writerTarget.get()).getHandle();
+            DeleteNode delete = (DeleteNode) getOnlyElement(findPlanNodes(plan, DeleteNode.class));
+            TableScanNode tableScan = getDeleteTableScan(delete, plan);
+            TupleDomain<ColumnHandle> originalEnforcedConstraint = tableScan.getEnforcedConstraint();
+            TableLayoutResult layoutResult = metadata.getLayout(
+                    session,
+                    tableHandle,
+                    new Constraint<>(originalEnforcedConstraint),
+                    Optional.of(ImmutableSet.copyOf(tableScan.getAssignments().values())));
+
+            return Optional.of(layoutResult.getLayout().getNewTableHandle());
+        }
+        return Optional.empty();
+    }
+
+    private static <T extends PlanNode> Optional<T> findSinglePlanNode(StreamingSubPlan plan, Class<T> clazz)
+    {
+        List<PlanNode> allMatches = findPlanNodes(plan, clazz);
+        switch (allMatches.size()) {
+            case 0:
+                return Optional.empty();
+            case 1:
+                return Optional.of((T) getOnlyElement(allMatches));
```

Comment:
You can avoid the unchecked cast:

* `List<PlanNode> allMatches` -> `List<T> allMatches`
* In `findPlanNodes`: `<T extends PlanNode> List<PlanNode>`  -> `<T extends PlanNode> List<T>`, `.findAll())` -> `.<T>findAll())`